packages/database/src/cachers/EntityHierarchyCacher.js

```diff
@@ -40,105 +34,115 @@ export class EntityHierarchyCacher {
+    const relations = useEntityRelationLinks
+      ? await this.getRelationsViaEntityRelation(hierarchyId, parentIds)
+      : await this.getRelationsCanonically(parentIds);
+    const childIdToParentId = reduceToDictionary(relations, 'child_id', 'parent_id');
+    const childIdsToAncestorIds = Object.fromEntries(
+      Object.entries(childIdToParentId).map(([childId, parentId]) => [
+        childId,
+        [parentId, ...parentIdsToAncestorIds[parentId]],
+      ]),
+    );
+    return childIdsToAncestorIds;
   }
 
-  async checkIfNextGenerationIsCanonical(hierarchyId, entityIds) {
-    const hierarchyLinkCount = await this.models.entityRelation.count({
+  async countEntityRelationChildren(hierarchyId, entityIds) {
+    return this.models.entityRelation.count({
       parent_id: entityIds,
       entity_hierarchy_id: hierarchyId,
     });
-    // if no entity relation links, the next gen must use the canonical "parent_id" links
-    return hierarchyLinkCount === 0;
   }
 
-  async getNextGenerationCanonically(entityId) {
+  async countCanonicalChildren(entityIds) {
     const canonicalTypes = Object.values(ORG_UNIT_ENTITY_TYPES);
-    const children = await this.models.entity.find(
-      { parent_id: entityId, type: canonicalTypes },
-      { columns: ['id'] },
-    );
-    return children.map(c => c.id);
+    return this.models.entity.count({
+      parent_id: entityIds,
+      type: canonicalTypes,
+    });
   }
 
-  async getNextGenerationViaEntityRelation(hierarchyId, entityId) {
+  async getRelationsViaEntityRelation(hierarchyId, parentIds) {
     // get any matching alternative hierarchy relationships leading out of these parents
-    const hierarchyLinks = await this.models.entityRelation.find({
-      parent_id: entityId,
+    return this.models.entityRelation.find({
+      parent_id: parentIds,
       entity_hierarchy_id: hierarchyId,
     });
-    return hierarchyLinks.map(l => l.child_id);
+  }
+
+  async getRelationsCanonically(entityId) {
+    const canonicalTypes = Object.values(ORG_UNIT_ENTITY_TYPES);
+    const children = await this.models.entity.find(
+      { parent_id: entityId, type: canonicalTypes },
+      { columns: ['id', 'parent_id'] },
+    );
+    return children.map(c => ({ child_id: c.id, parent_id: c.parent_id }));
   }
 
   /**
    * Stores the generation of ancestor/descendant info in the database
    * @param {string} hierarchyId
-   * @param {Entity[]} entityIds
-   * @param {Entity[]} ancestorIds In order of generational distance, with immediate parent at index 0
+   * @param {Entity[]} childIdsToAncestorIds  Ids of the child entities as keys, with the ids of their
+   *                                          ancestors in order of generational distance, with immediate
+   *                                          parent at index 0
    */
-  async cacheGeneration(hierarchyId, entityIds, ancestorIds) {
+  async cacheGeneration(hierarchyId, childIdsToAncestorIds) {
```

Comment:
Based on a previous comment:

```suggestion
  async cacheGeneration(hierarchyId, childIdToAncestorIds) {
```